ec/src/pairing.rs

```diff
@@ -0,0 +1,281 @@
+use crate::*;
+use ark_ff::{CyclotomicField, One, PrimeField, SquareRootField, Zero};
+use ark_serialize::SerializationError;
+use ark_std::rand::{
+    distributions::{Distribution, Standard},
+    Rng,
+};
+use ark_std::{
+    fmt::{Display, Formatter, Result as FmtResult},
+    io::{Read, Write},
+};
+
+pub trait Pairing: Sized + 'static + Copy + Debug + Sync + Send + Eq {
+    /// This is the scalar field of `Self::G1` and `Self::G2`.
+    type ScalarField: PrimeField + SquareRootField;
+
+    /// An element of G1.
+    type G1: CurveGroup<ScalarField = Self::ScalarField> + MulAssign<Self::ScalarField>; // needed due to https://github.com/rust-lang/rust/issues/69640
+
+    /// An element of G1 that has been preprocessed for use in a pairing.
+    type G1Prepared: Default
+        + Clone
+        + Send
+        + Sync
+        + Debug
+        + From<Self::G1>
+        + From<<Self::G1 as Group>::NormalForm>;
+
+    /// An element of G2.
+    type G2: CurveGroup<ScalarField = Self::ScalarField> + MulAssign<Self::ScalarField>; // needed due to https://github.com/rust-lang/rust/issues/69640
+
+    /// An element of G2 that has been preprocessed for use in a pairing.
+    type G2Prepared: Default
+        + Clone
+        + Send
+        + Sync
+        + Debug
+        + From<Self::G2>
+        + From<<Self::G2 as Group>::NormalForm>;
+
+    /// The extension field that hosts the target group of the pairing.
+    type TargetField: CyclotomicField;
+
+    /// Perform a miller loop with some number of (G1, G2) pairs.
+    #[must_use]
+    fn miller_loop<'a>(
+        i: impl IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>,
+    ) -> MillerLoopOutput<Self>;
+
+    /// Perform final exponentiation of the result of a miller loop.
+    #[must_use]
+    fn final_exponentiation(mlo: MillerLoopOutput<Self>) -> Option<PairingOutput<Self>>;
+
+    /// Computes a product of pairings.
+    #[must_use]
+    fn product_of_pairings<'a>(
+        i: impl IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>,
+    ) -> PairingOutput<Self> {
+        Self::final_exponentiation(Self::miller_loop(i)).unwrap()
+    }
```

Comment:
Can we please use a pair of references instead of a reference to a pair?

```rust
impl IntoIterator<Item = (&'a Self::G1Prepared, &'a Self::G2Prepared)>
```

It's almost impossible to compose this API with others that involve calling the pairing algorithm more than once.

Reply:
Yes, that's a good suggestion.